server/src/main/java/io/crate/planner/optimizer/Rule.java

```diff
@@ -33,7 +35,12 @@
 
     Pattern<T> pattern();
 
-    LogicalPlan apply(T plan, Captures captures, TableStats tableStats, TransactionContext txnCtx, NodeContext nodeCtx);
+    LogicalPlan apply(T plan,
+                      Captures captures,
+                      TableStats tableStats,
+                      TransactionContext txnCtx,
+                      NodeContext nodeCtx,
+                      Function<LogicalPlan, LogicalPlan> resolveLogicalPlan);
```

Comment:
If the rule is used in the classic optimizer, we can just pass in `Function.identity. In the Iterative Optimizer we pass in the Group Reference Resolver. This also means when a rule is accessing the `sources` of a Plan the related sources needs to be resolved by the Group Reference Resolver first. This is not done yet, and the rules need to be adapted.